File: JavaSimpleBackend/src/main/java/com/atesti/service/NaruciteljiService.java
package com.atesti.service;

import com.atesti.dto.NaruciteljRequest;
import com.atesti.dto.NaruciteljiPageResponse;
import com.atesti.entity.Narucitelj;
import com.atesti.exception.BadRequestException;
import com.atesti.exception.ResourceNotFoundException;
import com.atesti.repository.NaruciteljRepository;
import jakarta.persistence.criteria.Predicate;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Service
@RequiredArgsConstructor
public class NaruciteljiService {

    private final NaruciteljRepository naruciteljRepository;

    private static final List<String> ALLOWED_SORT_FIELDS = Arrays.asList(
            "id", "name", "mjesto", "drzava", "postanskiBroj", "OIB", "email", "createdAt", "updatedAt"
    );

    public Object getAllNarucitelji(String search, Integer page, Integer pageSize,
                                   String sortBy, String sortOrder,
                                   String nameFilter, String mjestoFilter,
                                   String drzavaFilter, String postanskiBrojFilter,
                                   String oibFilter, String emailFilter) {

        boolean hasGridParams = page != null || pageSize != null || sortBy != null || sortOrder != null
                || nameFilter != null || mjestoFilter != null || drzavaFilter != null
                || postanskiBrojFilter != null || oibFilter != null || emailFilter != null;

        String finalSortBy = ALLOWED_SORT_FIELDS.contains(sortBy) ? sortBy : "name";
        String finalSortOrder = "DESC".equalsIgnoreCase(sortOrder) ? "DESC" : "ASC";
        Sort sort = Sort.by(Sort.Direction.fromString(finalSortOrder), finalSortBy);

        Specification<Narucitelj> spec = buildSpecification(search, nameFilter, mjestoFilter,
                drzavaFilter, postanskiBrojFilter, oibFilter, emailFilter);

        if (!hasGridParams) {
            List<Narucitelj> rows = naruciteljRepository.findAll(spec, sort);
            return rows;
        }

        int pageNum = Math.max(1, page != null ? page : 1);
        int pageSizeNum = pageSize != null ? Math.min(100, Math.max(1, pageSize)) : 10;

        Page<Narucitelj> result = naruciteljRepository.findAll(spec, PageRequest.of(pageNum - 1, pageSizeNum, sort));

        return NaruciteljiPageResponse.builder()
                .items(result.getContent())
                .page(pageNum)
                .pageSize(pageSizeNum)
                .totalItems(result.getTotalElements())
                .totalPages(Math.max(1, result.getTotalPages()))
                .sortBy(finalSortBy)
                .sortOrder(finalSortOrder)
                .build();
    }

    public Narucitelj getNaruciteljiById(Long id) {
        return naruciteljRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Narucitelj not found"));
    }

    public Narucitelj createNarucitelj(NaruciteljRequest request) {
        String nameValue = request.getNarucitelj() != null ? request.getNarucitelj() : request.getName();
        if (nameValue == null || nameValue.isBlank()) {
            throw new BadRequestException("Error creating narucitelj: name is required");
        }

        Narucitelj narucitelj = Narucitelj.builder()
                .name(nameValue)
                .adresa(request.getAdresa())
                .mjesto(request.getMjesto())
                .postanskiBroj(request.getPostanskiBroj())
                .drzava(request.getDrzava())
                .OIB(request.getOIB())
                .ziroRacun(request.getZiroRacun())
                .ostalo(request.getOstalo())
                .kontaktOsoba(request.getKontaktOsoba())
                .telefon(request.getTelefon())
                .mobitel(request.getMobitel())
                .fax(request.getFax())
                .email(request.getEmail())
                .location(request.getMjesto() != null ? request.getMjesto() : "")
                .comment(request.getComment())
                .build();

        return naruciteljRepository.save(narucitelj);
    }

    public Narucitelj updateNarucitelj(Long id, NaruciteljRequest request) {
        Narucitelj narucitelj = naruciteljRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Narucitelj not found"));

        String nameValue = request.getNarucitelj() != null ? request.getNarucitelj() : request.getName();
        if (nameValue != null) {
            narucitelj.setName(nameValue);
        }
        if (request.getAdresa() != null) narucitelj.setAdresa(request.getAdresa());
        if (request.getMjesto() != null) narucitelj.setMjesto(request.getMjesto());
        if (request.getPostanskiBroj() != null) narucitelj.setPostanskiBroj(request.getPostanskiBroj());
        if (request.getDrzava() != null) narucitelj.setDrzava(request.getDrzava());
        if (request.getOIB() != null) narucitelj.setOIB(request.getOIB());
        if (request.getZiroRacun() != null) narucitelj.setZiroRacun(request.getZiroRacun());
        if (request.getOstalo() != null) narucitelj.setOstalo(request.getOstalo());
        if (request.getKontaktOsoba() != null) narucitelj.setKontaktOsoba(request.getKontaktOsoba());
        if (request.getTelefon() != null) narucitelj.setTelefon(request.getTelefon());
        if (request.getMobitel() != null) narucitelj.setMobitel(request.getMobitel());
        if (request.getFax() != null) narucitelj.setFax(request.getFax());
        if (request.getEmail() != null) narucitelj.setEmail(request.getEmail());
        if (request.getLocation() != null) narucitelj.setLocation(request.getLocation());
        if (request.getComment() != null) narucitelj.setComment(request.getComment());

        return naruciteljRepository.save(narucitelj);
    }

    public Narucitelj deleteNarucitelj(Long id) {
        Narucitelj narucitelj = naruciteljRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Narucitelj not found"));

        naruciteljRepository.delete(narucitelj);
        return narucitelj;
    }

    private Specification<Narucitelj> buildSpecification(String search, String nameFilter,
                                                         String mjestoFilter, String drzavaFilter,
                                                         String postanskiBrojFilter, String oibFilter,
                                                         String emailFilter) {
        return (root, query, cb) -> {
            List<Predicate> predicates = new ArrayList<>();

            if (search != null && !search.isBlank()) {
                String pattern = "%" + search + "%";
                Predicate searchPredicate = cb.or(
                        cb.like(root.get("name"), pattern),
                        cb.like(root.get("mjesto"), pattern),
                        cb.like(root.get("adresa"), pattern),
                        cb.like(root.get("email"), pattern),
                        cb.like(root.get("OIB"), pattern)
                );
                predicates.add(searchPredicate);
            }

            if (nameFilter != null && !nameFilter.isBlank()) {
                predicates.add(cb.like(root.get("name"), "%" + nameFilter.trim() + "%"));
            }
            if (mjestoFilter != null && !mjestoFilter.isBlank()) {
                predicates.add(cb.like(root.get("mjesto"), "%" + mjestoFilter.trim() + "%"));
            }
            if (drzavaFilter != null && !drzavaFilter.isBlank()) {
                predicates.add(cb.like(root.get("drzava"), "%" + drzavaFilter.trim() + "%"));
            }
            if (postanskiBrojFilter != null && !postanskiBrojFilter.isBlank()) {
                predicates.add(cb.like(root.get("postanskiBroj"), "%" + postanskiBrojFilter.trim() + "%"));
            }
            if (oibFilter != null && !oibFilter.isBlank()) {
                predicates.add(cb.like(root.get("OIB"), "%" + oibFilter.trim() + "%"));
            }
            if (emailFilter != null && !emailFilter.isBlank()) {
                predicates.add(cb.like(root.get("email"), "%" + emailFilter.trim() + "%"));
            }

            return cb.and(predicates.toArray(new Predicate[0]));
        };
    }
}

File: backend/services/naruciteljiService.js
const db = require('../models');

class NaruciteljiService {
  parseNumber(value, fallback) {
    const parsed = parseInt(value, 10);
    return Number.isNaN(parsed) ? fallback : parsed;
  }

  buildSearchWhere(searchTerm, filters = {}) {
    const where = {};
    const { Op } = db.Sequelize;

    if (searchTerm) {
      where[Op.or] = [
        { name: { [Op.like]: `%${searchTerm}%` } },
        { mjesto: { [Op.like]: `%${searchTerm}%` } },
        { adresa: { [Op.like]: `%${searchTerm}%` } },
        { email: { [Op.like]: `%${searchTerm}%` } },
        { OIB: { [Op.like]: `%${searchTerm}%` } }
      ];
    }

    const filterableFields = ['name', 'mjesto', 'drzava', 'postanskiBroj', 'OIB', 'email'];
    for (const field of filterableFields) {
      const value = filters[field];
      if (typeof value === 'string' && value.trim().length > 0) {
        where[field] = { [Op.like]: `%${value.trim()}%` };
      }
    }

    return where;
  }

  resolveSorting(sortBy, sortOrder) {
    const allowedSortFields = [
      'id',
      'name',
      'mjesto',
      'drzava',
      'postanskiBroj',
      'OIB',
      'email',
      'createdAt',
      'updatedAt'
    ];

    const finalSortBy = allowedSortFields.includes(sortBy) ? sortBy : 'name';
    const normalizedSortOrder = String(sortOrder || 'ASC').toUpperCase();
    const finalSortOrder = normalizedSortOrder === 'DESC' ? 'DESC' : 'ASC';

    return [finalSortBy, finalSortOrder];
  }

  async getAllNarucitelji(options = {}) {
    try {
      const searchTerm = options.search || '';
      const shouldPaginate = options.page !== undefined || options.pageSize !== undefined;
      const page = Math.max(1, this.parseNumber(options.page, 1));
      const pageSize = shouldPaginate
        ? Math.min(100, Math.max(1, this.parseNumber(options.pageSize, 10)))
        : null;
      const offset = shouldPaginate ? (page - 1) * pageSize : undefined;
      const order = [this.resolveSorting(options.sortBy, options.sortOrder)];
      const where = this.buildSearchWhere(searchTerm, options.filters);

      if (!shouldPaginate) {
        const rows = await db.narucitelj.findAll({
          where,
          order,
          include: [{
            model: db.radni_nalog,
            as: 'radniNalozi',
            required: false
          }]
        });

        return {
          items: rows,
          page: 1,
          pageSize: rows.length,
          totalItems: rows.length,
          totalPages: 1,
          sortBy: order[0][0],
          sortOrder: order[0][1]
        };
      }

      const result = await db.narucitelj.findAndCountAll({
        where,
        limit: pageSize,
        offset,
        order,
        distinct: true,
        include: [{
          model: db.radni_nalog,
          as: 'radniNalozi',
          required: false
        }]
      });

      const totalItems = result.count;
      const totalPages = Math.max(1, Math.ceil(totalItems / pageSize));

      return {
        items: result.rows,
        page,
        pageSize,
        totalItems,
        totalPages,
        sortBy: order[0][0],
        sortOrder: order[0][1]
      };
    } catch (error) {
      throw new Error(`Error fetching narucitelji: ${error.message}`);
    }
  }

  async getNaruciteljiById(id) {
    try {
      const narucitelj = await db.narucitelj.findByPk(id, {
        include: [{
          model: db.radni_nalog,
          as: 'radniNalozi',
          required: false
        }]
      });

      if (!narucitelj) {
        throw new Error('Narucitelj not found');
      }

      return narucitelj;
    } catch (error) {
      throw new Error(`Error fetching narucitelj: ${error.message}`);
    }
  }

  async createNarucitelj(data) {
    try {
      const nameValue = data.narucitelj || data.name;
      return await db.narucitelj.create({
        name: nameValue,
        adresa: data.adresa || null,
        mjesto: data.mjesto || null,
        postanskiBroj: data.postanskiBroj || null,
        drzava: data.drzava || null,
        OIB: data.OIB || null,
        ziroRacun: data.ziroRacun || null,
        ostalo: data.ostalo || null,
        kontaktOsoba: data.kontaktOsoba || null,
        telefon: data.telefon || null,
        mobitel: data.mobitel || null,
        fax: data.fax || null,
        email: data.email || null,
        location: data.mjesto,
        comment: data.comment || null
      });
    } catch (error) {
      throw new Error(`Error creating narucitelj: ${error.message}`);
    }
  }

  async updateNarucitelj(id, data) {
    try {
      const narucitelj = await db.narucitelj.findByPk(id);

      if (!narucitelj) {
        throw new Error('Narucitelj not found');
      }

      const nameValue = data.narucitelj !== undefined ? data.narucitelj : data.name;
      return await narucitelj.update({
        name: nameValue || narucitelj.name,
        adresa: data.adresa !== undefined ? data.adresa : narucitelj.adresa,
        mjesto: data.mjesto !== undefined ? data.mjesto : narucitelj.mjesto,
        postanskiBroj: data.postanskiBroj !== undefined ? data.postanskiBroj : narucitelj.postanskiBroj,
        drzava: data.drzava !== undefined ? data.drzava : narucitelj.drzava,
        OIB: data.OIB !== undefined ? data.OIB : narucitelj.OIB,
        ziroRacun: data.ziroRacun !== undefined ? data.ziroRacun : narucitelj.ziroRacun,
        ostalo: data.ostalo !== undefined ? data.ostalo : narucitelj.ostalo,
        kontaktOsoba: data.kontaktOsoba !== undefined ? data.kontaktOsoba : narucitelj.kontaktOsoba,
        telefon: data.telefon !== undefined ? data.telefon : narucitelj.telefon,
        mobitel: data.mobitel !== undefined ? data.mobitel : narucitelj.mobitel,
        fax: data.fax !== undefined ? data.fax : narucitelj.fax,
        email: data.email !== undefined ? data.email : narucitelj.email,
        location: data.location || narucitelj.location,
        comment: data.comment !== undefined ? data.comment : narucitelj.comment
      });
    } catch (error) {
      throw new Error(`Error updating narucitelj: ${error.message}`);
    }
  }

  async deleteNarucitelj(id) {
    try {
      const narucitelj = await db.narucitelj.findByPk(id);

      if (!narucitelj) {
        throw new Error('Narucitelj not found');
      }

      await narucitelj.destroy();
      return narucitelj;
    } catch (error) {
      throw new Error(`Error deleting narucitelj: ${error.message}`);
    }
  }
}

module.exports = new NaruciteljiService();
